packages/api/src/external/commonwell/link/shared.ts

```diff
@@ -44,6 +47,50 @@ export function patientWithCWData(
   return patientWithCW;
 }
 
+async function shouldDowngradeLink(link: NetworkLink): Promise<boolean> {
+  const identifiers = link.patient?.identifier || [];
+  const checkPromises = identifiers.map(async id => {
+    const idSystem = id.system?.replace(/^urn:oid:/, "");
+    if (idSystem) {
+      return (await getCQDirectoryEntryById(idSystem)) !== undefined;
+    }
+    return false;
+  });
+  const results = await Promise.all(checkPromises);
+  return results.includes(true); // true if any of the checks are true
+}
+
+async function assessIfCWCQDuplicatesAndDowngrade(
+  link: NetworkLink,
+  commonWell: CommonWellAPI,
+  queryMeta: RequestMetadata,
+  commonwellPatientId: string,
+  commonwellPersonId: string,
+  executionContext: string
+): Promise<NetworkLink | undefined> {
+  const shouldDowngrade = await shouldDowngradeLink(link);
+  if (shouldDowngrade && link._links?.downgrade?.href) {
+    console.log(`Downgrading link ${JSON.stringify(link, null, 2)} for ${commonwellPatientId}`);
+    return commonWell
+      .upgradeOrDowngradeNetworkLink(queryMeta, link._links.downgrade.href)
+      .catch(error => {
+        const msg = `Failed to downgrade link`;
+        console.log(`${msg}. Cause: ${error}`);
+        capture.message(msg, {
+          extra: {
+            commonwellPatientId,
+            commonwellPersonId,
+            cwReference: commonWell.lastReferenceHeader,
+            context: executionContext,
+          },
+          level: "error",
+        });
+        throw error;
+      });
+  }
```

Comment:
We should prob have an else here in case we decide to downgrade but there's no downgrade link, right? It's likely an edge case and maybe even a programming error, but still...

Reply:
handled